### PR TITLE
Create galore_introduction.md

### PR DESCRIPTION
**Pull request by Titus-von-Koeller**
Added the Galore blogpost to be published preferably today due to maximizing impact around GaLore and the integration into Transformers/TRL as part of todays Transformers release.

Thumbnail is still missing. I removed the line from the meta data for now, as I don't know what to put there.

```
thumbnail: /blog/assets/CHANGE_ME/thumbnail.png
```

### CODE REVIEW

**Review by younesbelkada — APPROVED** [review]
Thanks for this great blogpost !

File: _blog.yml [hunk 1]
```diff
@@ -3662,4 +3662,15 @@
     - hardware
     - nvidia
     - llm
+    - training
+  
+- local: galore_introduction
+  title: "GaLore: Advancing Large Model Training on Consumer-grade Hardware"
+  author: Titus-von-Koeller
+  thumbnail: /blog/assets/galore_introduction/thumbnail.png
+  date: March 19, 2024
```
Comment (younesbelkada) [inline _blog.yml:3671]:
```suggestion
  date: March 20, 2024
```

File: galore_introduction.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+The integration of GaLore into the training of large language models (LLMs) marks a significant advancement in the field of deep learning, particularly in terms of memory efficiency and the democratization of AI research. By allowing for the training of billion-parameter models on consumer-grade hardware, reducing memory footprint in optimizer states, and leveraging advanced projection matrix techniques, GaLore opens new horizons for researchers and practitioners with limited access to high-end computational resources.
+
+## Scaling LLMs with Consumer-Grade Hardware
+
+The capability of GaLore to facilitate the training of models with up to 7 billion parameters, such as those based on the LLaMA architecture, on consumer GPUs like the NVIDIA RTX 4090, is groundbreaking. This is achieved by significantly reducing the memory requirements traditionally associated with optimizer states and gradients during the training process. The approach leverages the inherent low-rank structure of gradients in deep neural networks, applying a projection that reduces the dimensionality of the data that needs to be stored and manipulated.
```
Comment (pcuenca) [inline galore_introduction.md:22]: (Just a comment: for Llama 2 meta changed the branding to just "Llama 2", with normal capitalization, so we could use Llama here if you happen to prefer it)

File: galore_introduction.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+## Memory Efficiency in Optimizer States
+
+The optimizer state, especially in adaptive optimization algorithms like Adam, represents a significant portion of the memory footprint during model training. GaLore addresses this by projecting the gradients into a lower-dimensional subspace before they are processed by the optimizer. This not only reduces the memory required to store these states but also maintains the effectiveness of the optimization process.
+
+The memory savings are substantial, with [the authors reporting](https://x.com/AnimaAnandkumar/status/1765613815146893348?s=20) “more than **82.5% reduction in memory for storing optimizer states during training**”, making it feasible to train larger models or use larger batch sizes within the same memory constraints.  reducing memory usage by up to 65.5% in optimizer states compared to traditional methods. When combined with 8-bit precision optimizers, these savings can be even more pronounced
```
Comment (pcuenca) [inline galore_introduction.md:28]:
```suggestion
The memory savings are substantial, with [the authors reporting](https://x.com/AnimaAnandkumar/status/1765613815146893348?s=20) “more than **82.5% reduction in memory for storing optimizer states during training**”, making it feasible to train larger models or use larger batch sizes within the same memory constraints. When combined with 8-bit precision optimizers, these savings can be even more pronounced.
```

File: galore_introduction.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+The ability to dynamically adjust these projections in response to changes in the gradient structure is a potent tool in the GaLore arsenal, allowing for more nuanced control over the memory-optimization trade-offs inherent in training large models.
+
+## Combining GaLore with 8-bit Optimizers
+
+The combination of GaLore with 8-bit precision optimizers represents a synergy that maximizes memory efficiency while maintaining the integrity and performance of the training process. 8-bit optimizers reduce the memory footprint by quantizing the optimizer states. When used in conjunction with GaLore's projection mechanism, the result is a highly memory-efficient training regime that does not compromise on model accuracy or convergence speed.
+
+This combination is particularly effective in scenarios where memory is a critical bottleneck, such as training large models on consumer-grade hardware or deploying models in memory-constrained environments. It enables the use of more complex models and larger datasets within the same hardware constraints, pushing the boundaries of what can be achieved with limited resources.
+
+## Implementation Details
```
Comment (pcuenca) [inline galore_introduction.md:42]:
```suggestion
## Implementation Details

```

File: galore_introduction.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+Integrating 8-bit optimizers with GaLore for training large language models (LLMs) involves quantizing the gradients, weights, and optimizer states to 8-bit representations. This quantization process significantly reduces the memory footprint, enabling the training of larger models or the use of larger batch sizes within the same memory constraints. The algorithmic details of this integration involve several key steps, some of which would benefit significantly from CUDA implementation for efficiency gains. GaLore opens new possibilities to integrate these techniques even closer with quantization and specialized parameterization of the matrices, which can lead to further reductions in memory usage and which we are currently working on in bitsandbytes.
```
Comment (pcuenca) [inline galore_introduction.md:43]:
```suggestion
Integrating 8-bit optimizers with GaLore for training large language models (LLMs) involves quantizing the gradients, weights, and optimizer states to 8-bit representations. This quantization process significantly reduces the memory footprint, enabling the training of larger models or the use of larger batch sizes within the same memory constraints. The algorithmic details of this integration involve several key steps, some of which would benefit significantly from native CUDA implementations for efficiency gains. GaLore opens new possibilities to integrate these techniques even more tightly with quantization and specialized parameterization of the matrices, which can lead to further reductions in memory usage. We are currently exploring this direction in the bitsandbytes library.
```

File: galore_introduction.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+## Implementation Details
+Integrating 8-bit optimizers with GaLore for training large language models (LLMs) involves quantizing the gradients, weights, and optimizer states to 8-bit representations. This quantization process significantly reduces the memory footprint, enabling the training of larger models or the use of larger batch sizes within the same memory constraints. The algorithmic details of this integration involve several key steps, some of which would benefit significantly from CUDA implementation for efficiency gains. GaLore opens new possibilities to integrate these techniques even closer with quantization and specialized parameterization of the matrices, which can lead to further reductions in memory usage and which we are currently working on in bitsandbytes. 
+
+### Algorithmic Overview of 8-bit Optimization with GaLore
```
Comment (pcuenca) [inline galore_introduction.md:45]:
```suggestion
### Algorithmic Overview of 8-bit Optimization with GaLore

```

File: galore_introduction.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+**Optimizer Update**: The 8-bit quantized gradients are used to update the model weights. This step involves de-quantizing the gradients back to floating-point format, applying the optimizer's update rule (e.g., Adam's moment update and parameter adjustment), and then quantizing the updated optimizer states back to 8-bit for storage.
+
+**De-quantization and Weight Update**: The 8-bit quantized weights undergo de-quantization to a floating-point representation for processing, albeit retaining the 8-bit precision inherent to their quantized form due to the limited range of values. This step is needed because standard operations in frameworks like PyTorch do not support 8-bit integers, and such integer weights cannot accommodate gradients. While this approach does not inherently enhance accuracy, it facilitates the practical application and gradient computation of quantized weights within the constraints of current deep learning libraries. Note that after de-quantization and before applying the weight update, GaLore employs one more projection that projects de-quantized low-rank updates back to the original space.
+
+## Use it with Hugging Face transformers
```
Comment (pcuenca) [inline galore_introduction.md:54]:
```suggestion
## Use it with Hugging Face Transformers
```

File: galore_introduction.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+To use GaLore optimizers with Hugging Face transformers library, you first need to update the transformers library to a version that supports GaLore optimizers, by either installing the latest update or installing transformers from source.
```
Comment (pcuenca) [inline galore_introduction.md:56]:
```suggestion
To use GaLore optimizers with the Hugging Face transformers library, you first need to update it to a version that supports GaLore optimizers, by either installing the latest update or installing transformers from source.
```

Comment (pcuenca) [inline galore_introduction.md:56]: I'd mention the transformers version here and maybe provide a pip install command.

File: galore_introduction.md [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+Then install the galore-torch library with `pip install galore-torch`. Below is a full working example of GaLore with transformers, for pretraining Mistral-7B on imdb dataset:
```
Comment (pcuenca) [inline galore_introduction.md:58]:
```suggestion
Then install the galore-torch library with `pip install galore-torch`. Below is a full working example of GaLore with transformers, for pretraining Mistral-7B on the imdb dataset:
```

File: galore_introduction.md [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+## Resources
+
+Please refer to [the original paper](https://arxiv.org/pdf/2403.03507.pdf). Twitter references: [1](https://twitter.com/AnimaAnandkumar/status/1765613815146893348) [2](https://x.com/_akhaliq/status/1765598376312152538?s=20) [3](https://x.com/tydsh/status/1765628222308491418?s=20). The paper also draws comparisons between GaLore and ReLoRA, which might be of interest to some readers. For readers with questions that remain unanswered, especially after review of the paper, or who would like to constructively discuss the results, please feel free to [join the author’s Slack community](https://galore-social.slack.com/join/shared_invite/zt-2ev152px0-DguuQ5WRTLQjtq2C88HBvQ#/shared-invite/email). For those interested in the further releases along these lines, please follow [Jiawei Zhao](https://twitter.com/jiawzhao) and [Titus von Koeller](https://twitter.com/Titus_vK) (for information on the latest `bitsandbytes` releases) as well as [Younes Belkada](https://twitter.com/younesbelkada) for the latest and greatest infos on quantization-related topics within and around the Hugging Face ecosystem.
```
Comment (pcuenca) [inline galore_introduction.md:108]:
```suggestion
Please refer to [the original paper](https://arxiv.org/pdf/2403.03507.pdf). Twitter references: [1](https://twitter.com/AnimaAnandkumar/status/1765613815146893348) [2](https://x.com/_akhaliq/status/1765598376312152538?s=20) [3](https://x.com/tydsh/status/1765628222308491418?s=20). The paper also draws comparisons between GaLore and ReLoRA, which might be of interest to some readers. For readers with questions that remain unanswered, especially after review of the paper, or who would like to constructively discuss the results, please feel free to [join the author’s Slack community](https://galore-social.slack.com/join/shared_invite/zt-2ev152px0-DguuQ5WRTLQjtq2C88HBvQ#/shared-invite/email). For those interested in further releases along these lines, please follow [Jiawei Zhao](https://twitter.com/jiawzhao) and [Titus von Koeller](https://twitter.com/Titus_vK) (for information on the latest `bitsandbytes` releases) as well as [Younes Belkada](https://twitter.com/younesbelkada) for the latest and greatest infos on quantization-related topics within and around the Hugging Face ecosystem.
```

**Review by pcuenca — APPROVED** [review]
Awesome 🔥

File: galore_introduction.md (hunk 2 above)
Comment (Titus-von-Koeller) [reply galore_introduction.md]: done

File: galore_introduction.md (hunk 8 above)
Comment (Titus-von-Koeller) [reply galore_introduction.md]: done

File: _blog.yml [hunk 11]
```diff
@@ -3662,4 +3662,15 @@
     - hardware
     - nvidia
     - llm
+    - training
+  
+- local: galore_introduction
```
[the remark this answers is not in the document]
Comment (Titus-von-Koeller) [reply _blog.yml]: done

**Comment by Titus-von-Koeller** [comment]
Ok, I integrated all your feedback. Thanks so much @pcuenca for the thorough review! Should be ready to merge now..

File: galore.md [hunk 12]
```diff
@@ -0,0 +1,112 @@
+---
+title: "GaLore: Advancing Large Model Training on Consumer-grade Hardware"
+authors:
+- user: Titus-von-Koeller
+- user: jiaweizhao
+  guest: true
+- user: matthewdouglas
```
Comment (matthewdouglas) [inline galore.md:7]:
```suggestion
- user: mdouglas
```
`matthewdouglas` is me too, but I don't use that account. One day I'll find out how I can merge them together :)

**Comment by Titus-von-Koeller** [comment]
Just merged. Thanks, everyone involved ❤️ Really happy about this!

**Pull request merged**